### PR TITLE
 Close #67. Adds additional test for code coverage metrics to be 100.

### LINKED ISSUE #67: Add additional code coverage 🎯 1/2 day
The code coverage levels are currently very low. Add additional tests to increase the code coverage level to as close to 100% as reasonably makes sense.

Files that are part of the `create-react-app` process (like the service worker) should be ignored by the code coverage reports.

Once the tests have been added, the "enforced" code coverage percentage should be updated to the new value such that if additional features are added without test coverage, the build fails.

### PR DESCRIPTION
**Pull request by kaxcode**

### CODE REVIEW

File: src/Alert.test.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import Enzyme from 'enzyme';
+import Alert from './Alert';
+import { mount } from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+
+Enzyme.configure({ adapter: new Adapter() });
+
+describe('Timer', () => {
```
Comment (dlindahl) [inline src/Alert.test.js]: This should be `Alert`

File: src/Alert.test.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  it('calls handleAlert when componentDidMount', () => {
```
Comment (dlindahl) [inline src/Alert.test.js]: I'd rewrite this as:

```js
describe('Alert', () => {
  describe('#componentDidMount', () => {
    it('calls #handleAlert')
  });
});
```

This namespaces all `componentDidMount` test cases as well as indicates that `handleAlert` is an instance method on `Alert` (cuz of `#` prefix)

Comment (dlindahl) [reply src/Alert.test.js]: Also, this is missing a test scenario for when the component is mounted and `show:false`

```js
it('calls #handleAlert when shown')
it('does not call #handleAlert when not shown')
```

Something like that.

File: src/Alert.test.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    // Arrange
+    function onDismissMock() {}
+    const wrapper = mount(
+      <Alert onDismiss={onDismissMock} msg="Alert!" show />
```
Comment (dlindahl) [inline src/Alert.test.js]: You can get rid of the anonymous function one of two ways:

```jsx
mount(<Alert onDismiss={jest.fn()} msg="" show/>);
```

or add [`empty`](https://www.npmjs.com/package/empty) as a dev dependency:

```jsx
import emptyFn from 'empty/function';

mount(<Alert onDismiss={emptyFn} msg="" show />
```

File: src/Alert.test.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    // Act
+    wrapper.instance().handleAlert();
+
+    // Assert
+    expect(wrapper.props().msg).toEqual('Alert!');
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('calls handleAlert when compoenentDidUpdate', () => {
```
Comment (dlindahl) [inline src/Alert.test.js]: Refactor to:

```js
describe('Alert', () => {
  describe('#componentDidUpdate', () => {
    it('calls #handleAlert when show changes from TRUE to FALSE')
    it('does not call #handleAlert when show changes from FALSE to TRUE')
  });
});

File: src/Stopwatch.test.js [hunk 5]
```diff
@@ -6,23 +6,99 @@ import Stopwatch from './Stopwatch';
...
+      startButton.simulate('click');
+
+      // Assert
+      expect(wrapper.state().clicked).toBe(true);
+      expect(setInterval).toHaveBeenCalledTimes(1);
+      expect(clearInterval).toHaveBeenCalledTimes(1);
+      expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 10);
+    });
+
+    it('it cannot be click again', () => {
```
Comment (dlindahl) [inline src/Stopwatch.test.js]: I'd rename to `it does nothing if already running`

Because it _can_ be clicked again, but only when its stopped.

File: src/Timer.test.js [hunk 6]
```diff
@@ -108,6 +112,18 @@ describe('Timer', () => {
         expect(wrapper.state().timeRemaining).toEqual(60 * minutes);
         expect(wrapper.state().timeRemaining).not.toBe(61 * minutes);
       });
+
+      it('returns when it reaches max minutes', () => {
```
Comment (dlindahl) [inline src/Timer.test.js]: Rename these to `it('does nothing when remaining time is at maximum')`, etc.

The reason being that all functions eventually return. We want to explicitly state that given a certain initial state (max/min time), the function is a no-op.

File: src/Timer.test.js [hunk 7]
```diff
@@ -195,53 +211,87 @@ describe('Timer', () => {
...
+      const clearButton = wrapper.find('#timer-clear').first();
       wrapper.setState({ timeRemaining: 19 * hours });
 
       // Act
       clearButton.simulate('click');
 
       // Assert
       expect(wrapper.state().timeRemaining).toEqual(0);
+      expect(wrapper.state().startClicked).toBe(false);
       expect(wrapper.state().timeRemaining).not.toBe(23 * hours);
```
Comment (dlindahl) [inline src/Timer.test.js]: This line is redundant in that the test has already asserted that `timeRemaining` is `0`, is there is no way for it to _not be_ 23hrs.

File: src/utils/humanizeTimer.js [hunk 8]
```diff
@@ -17,7 +17,7 @@ export function secondsToHour(time) {
 }
 
 export function secondsToMinutes(time) {
-  return Math.floor(time / 60 % 60);
+  return Math.floor((time / 60) % 60);
```
Comment (dlindahl) [inline src/utils/humanizeTimer.js]: This looks like a change that Prettier performed. You should put this into a separate commit.

File: src/Alert.test.js (hunk 4 above)
Comment (kaxcode) [reply src/Alert.test.js]: I think you mean 'does not call #handleAlert when show changes from TRUE to False` Show has to be true to call #handleAlert

Comment (dlindahl) [reply src/Alert.test.js]: 😬 oops. yes.

false => true `#handleAlert`
true => false `no-op`

Though now that I look at it, `#componentDidMount` from `true` to `true` should also be tested.

Comment (kaxcode) [reply src/Alert.test.js]: I added a test for true->true.

**Comment by dlindahl** [comment]
These two commit messages do not fit our formatting style. Also, the order of the two commits is backwards. Try this:

```
abc123 Fix formatting as recommended by prettier
678xyz Close #67 - Add additional tests to bring code coverage to 100%
```

* [x] Use the imperative mood in the subject line
* [x] Do not end the subject line with a period
* [x] Verb usage
* [x] Change is simple so no further context is necessary

Once those two changes are made, this is good to go. Good job! 🎉
